lib/glific/providers/gupshup/api_client.ex

```diff
@@ -2,12 +2,139 @@ defmodule Glific.Providers.Gupshup.ApiClient do
   @moduledoc """
   Http API client to intract with Gupshup
   """
+  alias Glific.Partners
   alias Plug.Conn.Query
 
+  @gupshup_url "https://api.gupshup.io/sm/api/v1"
+
   use Tesla
   # you can add , log_level: :debug to the below if you want debugging info
   plug Tesla.Middleware.Logger
 
   plug Tesla.Middleware.FormUrlencoded,
     encode: &Query.encode/1
+
+  @doc """
+  Making Tesla get call and adding api key in header
+  """
+  @spec gupshup_get(String.t(), String.t()) :: Tesla.Env.result()
+  def gupshup_get(url, api_key), do: get(url, headers: [{"apikey", api_key}])
+
+  @doc """
+  Making Tesla post call and adding api key in header
+  """
+  @spec gupshup_post(String.t(), any(), String.t()) :: Tesla.Env.result()
+  def gupshup_post(url, payload, api_key), do: post(url, payload, headers: [{"apikey", api_key}])
+
+  @spec get_credentials(non_neg_integer()) :: {:error, String.t()} | {:ok, map()}
+  defp get_credentials(org_id) do
+    organization = Partners.organization(org_id)
+
+    if is_nil(organization.services["bsp"]) do
+      {:error, "No active BSP available"}
+    else
+      bsp_credentials = organization.services["bsp"]
+      api_key = bsp_credentials.secrets["api_key"]
+      app_name = bsp_credentials.secrets["app_name"]
+
+      {:ok, %{api_key: api_key, app_name: app_name}}
+    end
+  end
+
+  @doc """
+  Fetching HSM templates for an organization
+  """
+  @spec get_templates(non_neg_integer()) :: Tesla.Env.result() | {:error, String.t()}
+  def get_templates(org_id) do
+    get_credentials(org_id)
+    |> case do
+      {:ok, credentials} ->
+        template_url = @gupshup_url <> "/template/list/" <> credentials.app_name
+        gupshup_get(template_url, credentials.api_key)
+
+      _ ->
+        :error
+    end
+  end
+
+  @doc """
+  Submitting HSM template for approval
+  """
+  @spec submit_template_for_approval(non_neg_integer(), map()) ::
+          Tesla.Env.result() | {:error, any()}
+  def submit_template_for_approval(org_id, payload) do
+    get_credentials(org_id)
+    |> case do
+      {:ok, credentials} ->
+        template_url = @gupshup_url <> "/template/add/" <> credentials.app_name
+        gupshup_post(template_url, payload, credentials.api_key)
+
+      _ ->
+        {:error, "error"}
```

Comment:
make it

{:error, "Please check your credential settings and ensure you have added the API Key and App Name also"}

use the above error across the entire file for credential failure